openhtf/util/threads.py

```diff
@@ -95,7 +95,7 @@ def async_raise(self, exc_type):
     assert self.is_alive(), 'Only running threads have a thread identity'
     result = ctypes.pythonapi.PyThreadState_SetAsyncExc(
         ctypes.c_long(self.ident), ctypes.py_object(exc_type))
-    if result == 0:
+    if result == 0 and self.is_alive():
       raise ValueError('Thread ID was invalid.', self.ident)
```

Comment:
Some interesting notes in the documentation of `PyThreadState_SetAsyncExc`: 

> To prevent naive misuse, you must write your own C extension to call this. Must be called with the GIL held.

I wonder what's the worst that could happen? Your change already seems like an improvement--I can't tell in what situation this ValueError would ever be raised anymore. The only loophole I've thought of is that since thread identifiers can be “recycled,” I wonder if there's a very small chance that we end up killing some arbitrary other thread that was not meant to be killed.

Reply:
Yeah, I think the C extension that holds the GIL is the only way to completely eliminate possibility of a race condition here.  I don't want to take that on now though but did add a note in the class docstring.

From documentation of threading.Thread on ident:

> The ‘thread identifier’ of this thread or None if the thread has not been started. This is a nonzero integer. See the thread.get_ident() function. Thread identifiers may be recycled when a thread exits and another thread is created. The identifier is available even after the thread has exited.

So the assertion message is a little misleading.  Since it's not running threads but started threads that have an ident.  I changed the message.

I also put a check on is_alive before calling PyThreadState_SetAsyncExc.  This might be a bit of overkill considering we also check on result == 0 and self.is_alive(), but it seems to me that it would be better to not call at all if the thread is not alive.

Reply:
The GIL is held when running Python code, if you want to ensure that it's held throughout a function, you can temporarily set the check interval (sys.setcheckinterval) to a very large number. That will ensure that Python doesn't even attempt to switch out the GIL, since the check interval is how often the interpreter checks things, like if other threads are ready to run.

https://docs.python.org/2/c-api/init.html#thread-state-and-the-global-interpreter-lock

Reply:
As far as I can tell this solution should work. But I know there are also some concerns about this being a hack. This article, [A Neat Python Hack? No, Broken Code](http://www.dr-josiah.com/2011/07/neat-python-hack-no-broken-code.html), points out that code in the critical section may be releasing the GIL unbeknownst to you. This doesn't *seem* to be an issue in our case, if we're completely sure about what `ctypes.pythonapi.PyThreadState_SetAsyncExc` is doing (which I'm not).

Pending an airtight solution, I'm fine with @kdsudac's existing improvement, which I believe leaves only the remote, probably-never-gonna-happen possibility of killing the wrong thread